### PR TITLE
Mk nikshay audit cookie safe

### PR DESCRIPTION
**Pull request by mkangia**
Changes for UATBC security audit

Make CSRF cookie http only, [doc](https://docs.google.com/document/d/12_38B4LFhdzz7XSGDOKxoreBLis6th_O9r5yxTS18ZE/edit#bookmark=id.xbasi7d0m6jj)

[Test plan](https://docs.google.com/document/d/1h7ztnuJhIr_9DABqYju1G3857kE9yCCEX7c9GLWylQg/edit)

### CODE REVIEW

File: corehq/apps/hqwebapp/static/hqwebapp/js/ajax_csrf_setup.js
Comment (dimagimon) [inline corehq/apps/hqwebapp/static/hqwebapp/js/ajax_csrf_setup.js]: '$csrf_token' is not defined. (no-undef)

File: corehq/apps/app_manager/templates/app_manager/v1/summary.html
Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/summary.html]: @orangejenny / @biyeun Can you help understand what i am missing here? I have done the change for v2 summary as well and using `build_app_manager_v2_diffs` generates no diff files. But the build fails: https://travis-ci.org/dimagi/commcare-hq/jobs/215294313

Comment (snopoke) [inline corehq/apps/app_manager/templates/app_manager/v1/summary.html]: pretty sure there would be some diff changes - even if they are only line numbers

Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/summary.html]: Updated the branch with master and ran `build_app_manager_v2_diffs` again. It generated a diff file `corehq/apps/app_manager/tests/data/v2_diffs/templates/summary.html.diff.txt`.

Comment (orangejenny) [inline corehq/apps/app_manager/templates/app_manager/v1/summary.html]: That sounds right (merging master) - apologies for not having gotten to look at this, but glad you figured it out.

Comment (mkangia) [inline corehq/apps/app_manager/templates/app_manager/v1/summary.html]: Merged master to update to new django version. Then just worked. :)

**Review by snopoke — APPROVED** [review]

**Comment by snopoke** [comment]
I assume you'll still put this on staging to test it.

**Comment by mkangia** [comment]
@snopoke 

> I assume you'll still put this on staging to test it.

No doubt about it.

**Comment by mkangia** [comment]
This is on staging now!

**Comment by snopoke** [comment]
@mkangia can you make a QA ticket and list all the places that could be affected by this.

**Comment by mkangia** [comment]
@snopoke 
Added test plan 
https://docs.google.com/document/d/1h7ztnuJhIr_9DABqYju1G3857kE9yCCEX7c9GLWylQg/edit

I thought it was better to do a check by actually seeing it break and then work with the change. I have added all the places i thought would be affected and also did test some of them to see them break by reverting the change and then work with the change. I don't think its possible to point out each and every place that can get affected and test it since there are a couple of generic js files that are used at quite some places.
The only place i see this can break if the template/view being rendered does not extend from `style/base.html` which adds the DOM element. Otherwise i think this should be good.
Should this still go through QA?

Moreover, i am a little surprised that no test cases failed.

**Comment by emord** [comment]
Had to merge master into this branch to deploy staging. There were conflicts due to javascript externalize prs

@mkangia @orangejenny could you review the merge commit?

**Comment by dimagimon** [comment]
There are 19 errors:

* corehq/apps/app_manager/static/app_manager/js/vellum_beta/src/jstree.js, line 1 - 'define' is not defined. (no-undef)
* corehq/apps/app_manager/static/app_manager/js/vellum_beta/src/jstree.js, line 8 - 'util' is defined but never used (no-unused-vars)
* corehq/apps/accounting/tests/generator.py, line 32 - F401 'WebUser' imported but unused
* corehq/apps/app_manager/suite_xml/features/mobile_ucr.py, line 200 - E501 line too long (122 > 115 characters)
* corehq/apps/app_manager/tasks.py, line 15 - E226 missing whitespace around arithmetic operator
* corehq/apps/app_manager/views/modules.py, line 970 - F812 list comprehension redefines 'module' from line 946
* corehq/apps/change_feed/consumer/feed.py, line 150 - E501 line too long (116 > 115 characters)
* corehq/apps/cleanup/management/commands/reprocess_error_forms.py, line 112 - E502 the backslash is redundant between brackets
* corehq/apps/cleanup/management/commands/reprocess_error_forms.py, line 115 - E502 the backslash is redundant between brackets
* corehq/apps/domain/models.py, line 428 - E226 missing whitespace around arithmetic operator
* corehq/apps/domain/models.py, line 548 - E226 missing whitespace around arithmetic operator
* corehq/apps/hqadmin/views.py, line 43 - F401 'send_html_email_async' imported but unused
* corehq/apps/hqadmin/views.py, line 64 - F401 'get_app_ids_in_domain' imported but unused
* corehq/apps/hqadmin/views.py, line 80 - F401 'force_to_date' imported but unused
* corehq/apps/locations/forms.py, line 23 - F401 'raw_username' imported but unused
* corehq/apps/users/bulkupload.py, line 23 - F401 'submit_mapping_case_block' imported but unused
* corehq/apps/users/middleware.py, line 8 - E303 too many blank lines (3)
* corehq/apps/users/middleware.py, line 13 - W291 trailing whitespace
* corehq/apps/users/views/mobile/users.py, line 56 - F401 'is_location_safe' imported but unused

**Comment by orangejenny** [comment]
@emord @mkangia Externalizing js doesn't handle csrf tokens yet. I reverted the merge commit and pulled this branch off staging. I should be able to get the csrf thing figured out and this branch back on staging in a day or two - Manish, let me know if there's urgency (I'm guessing this branch is just being passively tested on staging).

**Comment by dimagimon** [comment]
There are 19 errors:

* corehq/apps/app_manager/static/app_manager/js/vellum_beta/src/jstree.js, line 1 - 'define' is not defined. (no-undef)
* corehq/apps/app_manager/static/app_manager/js/vellum_beta/src/jstree.js, line 8 - 'util' is defined but never used (no-unused-vars)
* corehq/apps/accounting/tests/generator.py, line 32 - F401 'WebUser' imported but unused
* corehq/apps/app_manager/suite_xml/features/mobile_ucr.py, line 200 - E501 line too long (122 > 115 characters)
* corehq/apps/app_manager/tasks.py, line 15 - E226 missing whitespace around arithmetic operator
* corehq/apps/app_manager/views/modules.py, line 970 - F812 list comprehension redefines 'module' from line 946
* corehq/apps/change_feed/consumer/feed.py, line 150 - E501 line too long (116 > 115 characters)
* corehq/apps/cleanup/management/commands/reprocess_error_forms.py, line 112 - E502 the backslash is redundant between brackets
* corehq/apps/cleanup/management/commands/reprocess_error_forms.py, line 115 - E502 the backslash is redundant between brackets
* corehq/apps/domain/models.py, line 428 - E226 missing whitespace around arithmetic operator
* corehq/apps/domain/models.py, line 548 - E226 missing whitespace around arithmetic operator
* corehq/apps/hqadmin/views.py, line 43 - F401 'send_html_email_async' imported but unused
* corehq/apps/hqadmin/views.py, line 64 - F401 'get_app_ids_in_domain' imported but unused
* corehq/apps/hqadmin/views.py, line 80 - F401 'force_to_date' imported but unused
* corehq/apps/locations/forms.py, line 23 - F401 'raw_username' imported but unused
* corehq/apps/users/bulkupload.py, line 23 - F401 'submit_mapping_case_block' imported but unused
* corehq/apps/users/middleware.py, line 8 - E303 too many blank lines (3)
* corehq/apps/users/middleware.py, line 13 - W291 trailing whitespace
* corehq/apps/users/views/mobile/users.py, line 56 - F401 'is_location_safe' imported but unused

**Comment by orangejenny** [comment]
@mkangia Figured out how to make csrf_token work with initial_page_data and merged master. 

But...I suspect https://github.com/dimagi/commcare-hq/commit/89b0e9b1dd63643b1b3aade0f205adc385b55d72 in this branch is going to result in more merge conflicts (or silent bad merges). You've been adding csrf tokens to our angular pages at the same time that I've been pulling the inline javascript out of those pages, into separate files.

There are some pages where I've recently split inline JavaScript out into separate files but this branch doesn't reflect those changes, even after merging master. For an example, [summary.html on master](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/templates/app_manager/v1/summary.html) no longer uses a `js-inline` block, but [summary.html on this branch](https://github.com/dimagi/commcare-hq/blob/mk-nikshay-audit-cookie-safe/corehq/apps/app_manager/templates/app_manager/v1/summary.html) still has the `js-inline` block. `summary.html` (both v1 and v2), `dashboard_domain.html`, and `web_users.html` are all like this, and `export_list.html` will be once https://github.com/dimagi/commcare-hq/pull/15730 is merged.

For these pages, the csrf token will need to be added to initial page data so the external js file can access it. This turns out to be straightforward. Using the report builder type selection page as an example:
1) [The view](https://github.com/dimagi/commcare-hq/blob/mk-nikshay-audit-cookie-safe/corehq/apps/userreports/views.py#L381) needs to add the csrf token to the view context
2) [The template](https://github.com/dimagi/commcare-hq/blob/mk-nikshay-audit-cookie-safe/corehq/apps/userreports/templates/userreports/reportbuilder/report_type_select.html#L25) needs to add the csrf token to initial page data
3) [The js file](https://github.com/dimagi/commcare-hq/blob/mk-nikshay-audit-cookie-safe/corehq/apps/userreports/static/userreports/js/report_type_select.js#L15) can then grab the token from initial page data.

**Comment by orangejenny** [comment]
@mkangia ...never mind that, I just found you've already got [a pattern](https://github.com/dimagi/commcare-hq/blob/mk-nikshay-audit-cookie-safe/corehq/apps/export/static/export/js/download_export.ng.js#L18) for getting at the csrf token in js files.

**Comment by mkangia** [comment]
sorry @emord you had to take a shot at resolving the conflicts. Those were confusing.

Thanks @orangejenny for looking into this. I reverted the branch to before any merges and then merged master to resolve conflicts. I think its as desired now.

FYI:
I have been following a simple rule:
1. for js files fetch csrftoken from the DOM element (to avoid issues in static files which would get precompiled)
2. for html inlines use django csrftoken method (this is done instead of picking from DOM just to be extra careful in case inline JS is called before DOM element is ready)

**Comment by mkangia** [comment]
@snopoke 

> @snopoke
Added test plan
https://docs.google.com/document/d/1h7ztnuJhIr_9DABqYju1G3857kE9yCCEX7c9GLWylQg/edit
I thought it was better to do a check by actually seeing it break and then work with the change. I have added all the places i thought would be affected and also did test some of them to see them break by reverting the change and then work with the change. I don't think its possible to point out each and every place that can get affected and test it since there are a couple of generic js files that are used at quite some places.
The only place i see this can break if the template/view being rendered does not extend from style/base.html which adds the DOM element. Otherwise i think this should be good.
Should this still go through QA?

**Comment by snopoke** [comment]
If you're happy with it and have tested all the things you can then that's ok with me

**Comment by mkangia** [comment]
@orangejenny Seems like this is good to go live now. I have done a second round of testing post updates and merges with master. Do you want to give it another review?

**Comment by orangejenny** [comment]
Looks good to me.

If you have the time, it would be nice if the inline javascript used the HTML container style instead of the django tag, because I'm in the process of moving a lot of our JavaScript to external files and will have to make that change later. Doesn't have to block this PR, though.

**Comment by orangejenny** [comment]
Newest commits look good, thank you for getting the django `{{ csrf_token }}` out of the inline javascript.

File: custom/_legacy/pact/templates/pact/patient/pactpatient_providers.html
Comment (sravfeyn) [inline custom/_legacy/pact/templates/pact/patient/pactpatient_providers.html]: I haven't checked for rest of the templates that this is changed, but this template extends `"reports/async/default.html"` which doesn't extend from `style/base.html`, so will it not miss the element `csrfTokenContainer `?

Comment (mkangia) [inline custom/_legacy/pact/templates/pact/patient/pactpatient_providers.html]: Agreed that that is what it looks like from code. To confirm, I checked for the view that renders it [here](https://www.commcarehq.org/a/pact/reports/custom/patient/?patient_id=0bd56681dce2b7a5140f75266d20880e&view=providers) and it had the container.

**Pull request merged**
